Pioneer send support

kpalczewski opened the pull request:
pioneer receivers use NEC codes for IR communication.
some codes are 2 NEC codes separated by a long pauze. this PR adds support for sending those codes.

you can find Pioneer codes here:
https://www.pioneerelectronics.com/PUSA/Support/Home-Entertainment-Custom-Install/IR+Codes/A+V+Receivers

kpalczewski committed:
add Support for sending Piooneer 64bit codes.

kpalczewski committed:
Revert "add Support for sending Piooneer 64bit codes."

This reverts commit 8cc8af2ff87c2ea4876496100362ce787a022958.

kpalczewski committed:
add support for Pioneer 64bit codes.

kpalczewski committed:
updated code for failed checks.

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+     kNECBits * (kNecBitMarkTicks + kNecOneSpaceTicks) +
+     kNecBitMarkTicks);
+
+#if (SEND_PIONEER)
```
Nitpick:
```
#if SEND_PIONEER
```

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+     kNecBitMarkTicks);
+
+#if (SEND_PIONEER)
+// Send a raw NEC(Renesas) formatted message.
```
Fix this comment. It's not NEC. ;-)

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+//
+// Args:
+//   data:   The message to be sent.
+//   nbits:  The number of bits of the message to be sent. Typically kNECBits.
```
Fix the comment. Not kNECBits.

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+// Status: STABLE / Known working.
+//
+// Ref:
+//  http://www.sbprojects.com/knowledge/ir/nec.php
```
Ref is for NEC. Pls fix.

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+  // prepare codes
+  uint64_t NECcode1 = data;
+  uint64_t NECcode2;
+  NECcode1 >>= 32;  // 1st code
```
This should probably be "nbits / 2", not 32.

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+const uint16_t kNecOneSpace = kNecOneSpaceTicks * kNecTick;
+const uint16_t kNecZeroSpaceTicks = 1;
+const uint16_t kNecZeroSpace = kNecZeroSpaceTicks * kNecTick;
+const uint16_t kPioneerCodeSeparatorTicks = 989;
```
It's not ticks. Should just be `kPioneerCodeSeparator`

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+  uint64_t NECcode1 = data;
+  uint64_t NECcode2;
+  NECcode1 >>= 32;  // 1st code
+  NECcode2 = data & 0xffffffffUL;  // 2nd code
```
And probably should calculate this off the nbits size, rather than hard code it. That, or add a test at the start to ensure it's always 64 (nbits).

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+  NECcode2 = data & 0xffffffffUL;  // 2nd code
+
+  // send 1st NEC code
+  sendGeneric(kNecHdrMark, kNecHdrSpace,
```
Perhaps just call `sendNEC(NECcode1, nbits/2);` instead?

crankyoldgit replied:
I think you can use:
`sendNEC(PioneerCode1, nbits / 2);` instead of sendGeneric() here.

kpalczewski replied:
sure. I'll test this later.

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+                38, true, repeat - 1,  // We've already sent a one message.
+                33);
+}
+#endif
```
`#endif  // SEND_PIONEER`

crankyoldgit commented on src/IRremoteESP8266.h:
```diff
@@ -193,6 +193,8 @@
 #define DECODE_PANASONIC_AC    true
 #define SEND_PANASONIC_AC      true
 
+#define SEND_PIONEER           true
```
Please add:
```
#define DECODE_PIONEER false  // Not yet implemented.
```

crankyoldgit commented on src/IRremoteESP8266.h:
```diff
@@ -267,6 +269,7 @@ enum decode_type_t {
   LUTRON,
   ELECTRA_AC,
   PANASONIC_AC,
+  PIONEER,  // Technically not a protocol, but an encoding.
```
I'd argue it is. Remove the comment. ;-)

crankyoldgit commented on src/IRremoteESP8266.h:
```diff
@@ -420,6 +424,7 @@ const uint16_t kWhynterBits = 32;
 #define TOSHIBA_AC_STATE_LENGTH       kToshibaACStateLength
 #define TROTEC_COMMAND_LENGTH         kTrotecStateLength
 #define WHYNTER_BITS                  kWhynterBits
+#define PIONEER_BITS                  kPioneerBits
```
No need to add this. That's only for legacy stuff. >2.5.0 We don't use `#defines` for it these.

crankyoldgit commented on src/IRsend.h:
```diff
@@ -299,6 +299,10 @@ void send(uint16_t type, uint64_t data, uint16_t nbits);
 #endif
+#if (SEND_PIONEER)
+  void sendPioneer(uint64_t data, uint16_t nbits = kPioneerBits,
+               uint16_t repeat = kNoRepeat);
```
Style nitpick:
Please add more spaces so that the 'uint64_t`s line up/start after the `(`.

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+#include <algorithm>
+#include "IRrecv.h"
+#include "IRsend.h"
+#include "IRutils.h"
```
Add:
```
#include "ir_NEC.h"
```

(Similar to the `ir_Sherwood.cpp` situation.)

kpalczewski replied:
there is no ir_NEC.h in the library. 
ir_Sherwood.cpp has included:
`#include <algorithm>`
`#include "IRsend.h`

crankyoldgit replied:
Oops. My bad. I thought there was. I might have to split it out and add it.

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+
+// Constants
+// Ref:
+//  http://www.sbprojects.com/knowledge/ir/nec.php
```
Not nec.

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+//                        P      I  O   O N  NN E    E    R R
+//                        P     III  OOO  N   N EEEE EEEE R  RR
+
+// NEC originally added from https://github.com/shirriff/Arduino-IRremote/
```
Not NEC.

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+// Ref:
+//  http://www.sbprojects.com/knowledge/ir/nec.php
+
+const uint16_t kNecTick = 560;
```
These (kNecEtc...) should all come/be included with the `#include "ir_NEC.h"` above

kpalczewski replied:
there is no `ir_NEC.h`, so I'm a little confused. Should like rewrite `ir_NEC.cpp`, so it would include variables from `ir_NEC.h` and `ir_Pioneer.cpp` should use those variables as well?

crankyoldgit replied:
Yeah. I'll probably need to do that. But for now, assume my comment about using `ir_NEC.h` for now is wrong/bad. I clearly had a brain fade. I'll split it out and clean this up later.

crankyoldgit commented on src/ir_Pioneer.cpp:
```diff
@@ -0,0 +1,101 @@
+              0, 0, 38, true, 0,  // Repeats are handled later.
+              33);
+
+  // send 2nd NEC code
```
Similarly you probably can replace the rest of the code/function from here on in with just:
```
  sendNEC(NECcode2, nbits / 2, repeat);
```

crankyoldgit reviewed (changes requested):
First off, Congratulations on adding a new protocol to the library. It's a big effort. Thanks!!

Sorry for the heaps of comments/suggestions/fixes etc.
I'm just trying to make it as good as we can.

Do you have any `rawData` lines which records what the protocol looks like? If so, I'll add a decoder routine, and some unit test code etc so we know we can be fairly sure it behaves like we want it to.

David

kpalczewski committed:
remove NEC references, clean up the code.

kpalczewski committed:
fix failed checks.

kpalczewski committed:
fix failed checks.

crankyoldgit committed:
Add Pioneer Decoding.
* Simplify the sendPioneer() routine.
* Add encodePioneer().
* Add decodePioneer().
* Unit tests for sendPioneer() & decodePioneer().
* Code style/lint cleanups.

crankyoldgit committed:
Minor code & comment cleanup.
No real change to functionality.

crankyoldgit committed:
Add Pioneer to tools/gc_decode.

crankyoldgit commented:
@kpalczewski I've cleaned up your code, and I verified that it produced the same output as your original PR. However, when I went to https://www.pioneerelectronics.com/PUSA/Support/Home-Entertainment-Custom-Install/IR+Codes/A+V+Receivers and looked at some of the PRONTO/HEX codes they have for the two-code messages for some devices. e.g. Spotify etc. They are not the same as what your/my code produce.

I think you may have incorrectly coded up how the Pioneer protocol works. Can you please capture a `rawData[]` line for one of these messages that you tried to emulate from the unmodified **2.5.1 release** version of the library's `IRrecvDumpV2` program? I want to be sure we are correctly emulating the real remote before we merge this into the library.

kpalczewski commented:
@crankyoldgit thank you for the cleanup.
Here are some RAW data for short and long codes:

VOL+ (short):
Encoding  : NEC
Code      : 659A50AF (32 bits)
Library   : v2.5.1

Raw Timing[67]:
   +  8554, -  4184,    +   592, -   472,    +   592, -  1524,    +   592, -  1524, 
   +   592, -   472,    +   592, -   472,    +   598, -  1520,    +   592, -   476, 
   +   594, -  1520,    +   596, -  1520,    +   598, -   472,    +   592, -   472, 
   +   592, -  1524,    +   592, -  1524,    +   594, -   472,    +   594, -  1524, 
   +   592, -   472,    +   594, -   476,    +   592, -  1520,    +   598, -   472, 
   +   592, -  1524,    +   594, -   470,    +   594, -   472,    +   596, -   472, 
   +   592, -   472,    +   592, -  1524,    +   592, -   472,    +   592, -  1524, 
   +   592, -   472,    +   594, -  1524,    +   598, -  1520,    +   594, -  1524, 
   +   594, -  1524,    +   594

uint16_t rawData[67] = {8554, 4184,  592, 472,  592, 1524,  592, 1524,  592, 472,  592, 472,  598, 1520,  592, 476,  594, 1520,  596, 1520,  598, 472,  592, 472,  592, 1524,  592, 1524,  594, 472,  594, 1524,  592, 472,  594, 476,  592, 1520,  598, 472,  592, 1524,  594, 470,  594, 472,  596, 472,  592, 472,  592, 1524,  592, 472,  592, 1524,  592, 472,  594, 1524,  598, 1520,  594, 1524,  594, 1524,  594};  // NEC 659A50AF
uint32_t address = 0xA6;
uint32_t command = 0xA;
uint64_t data = 0x659A50AF;

VOL- (short):
Encoding  : NEC
Code      : 659AD02F (32 bits)
Library   : v2.5.1

Raw Timing[67]:
   +  8554, -  4184,    +   592, -   472,    +   596, -  1520,    +   596, -  1520, 
   +   596, -   472,    +   592, -   472,    +   592, -  1524,    +   592, -   472, 
   +   596, -  1520,    +   596, -  1520,    +   592, -   476,    +   592, -   472, 
   +   592, -  1524,    +   592, -  1524,    +   592, -   472,    +   594, -  1524, 
   +   592, -   472,    +   596, -  1520,    +   596, -  1520,    +   598, -   470, 
   +   594, -  1520,    +   598, -   472,    +   594, -   470,    +   594, -   472, 
   +   596, -   472,    +   592, -   472,    +   592, -   472,    +   596, -  1520, 
   +   592, -   476,    +   594, -  1520,    +   596, -  1520,    +   596, -  1520, 
   +   598, -  1520,    +   598

uint16_t rawData[67] = {8554, 4184,  592, 472,  596, 1520,  596, 1520,  596, 472,  592, 472,  592, 1524,  592, 472,  596, 1520,  596, 1520,  592, 476,  592, 472,  592, 1524,  592, 1524,  592, 472,  594, 1524,  592, 472,  596, 1520,  596, 1520,  598, 470,  594, 1520,  598, 472,  594, 470,  594, 472,  596, 472,  592, 472,  592, 472,  596, 1520,  592, 476,  594, 1520,  596, 1520,  596, 1520,  598, 1520,  598};  // NEC 659AD02F
uint32_t address = 0xA6;
uint32_t command = 0xB;
uint64_t data = 0x659AD02F;

POWER (long):
Encoding  : NEC
Code      : 659A857A (32 bits)
Library   : v2.5.1

Raw Timing[135]:
   +  8550, -  4188,    +   592, -   472,    +   592, -  1524,    +   592, -  1524, 
   +   592, -   472,    +   594, -   470,    +   594, -  1524,    +   592, -   476, 
   +   592, -  1524,    +   594, -  1524,    +   592, -   472,    +   592, -   472, 
   +   592, -  1524,    +   592, -  1524,    +   592, -   472,    +   592, -  1524, 
   +   592, -   476,    +   592, -  1524,    +   592, -   472,    +   592, -   472, 
   +   592, -   476,    +   592, -   472,    +   592, -  1524,    +   592, -   472, 
   +   592, -  1524,    +   592, -   476,    +   588, -  1524,    +   592, -  1524, 
   +   592, -  1524,    +   592, -  1526,    +   592, -   476,    +   592, -  1524, 
   +   592, -   472,    +   594, - 25284,    +  8552, -  4184,    +   592, -  1524, 
   +   592, -  1524,    +   592, -  1524,    +   592, -  1524,    +   592, -   472, 
   +   592, -  1524,    +   592, -   472,    +   592, -  1524,    +   592, -   476, 
   +   592, -   472,    +   592, -   472,    +   594, -   476,    +   588, -  1528, 
   +   588, -   476,    +   592, -  1524,    +   594, -   470,    +   592, -   476, 
   +   588, -   476,    +   592, -  1524,    +   594, -  1524,    +   592, -  1526, 
   +   592, -  1524,    +   594, -   472,    +   592, -  1524,    +   592, -  1524, 
   +   592, -  1524,    +   592, -   472,    +   592, -   472,    +   592, -   476, 
   +   592, -   472,    +   592, -  1524,    +   592, -   472,    +   594

uint16_t rawData[135] = {8550, 4188,  592, 472,  592, 1524,  592, 1524,  592, 472,  594, 470,  594, 1524,  592, 476,  592, 1524,  594, 1524,  592, 472,  592, 472,  592, 1524,  592, 1524,  592, 472,  592, 1524,  592, 476,  592, 1524,  592, 472,  592, 472,  592, 476,  592, 472,  592, 1524,  592, 472,  592, 1524,  592, 476,  588, 1524,  592, 1524,  592, 1524,  592, 1526,  592, 476,  592, 1524,  592, 472,  594, 25284,  8552, 4184,  592, 1524,  592, 1524,  592, 1524,  592, 1524,  592, 472,  592, 1524,  592, 472,  592, 1524,  592, 476,  592, 472,  592, 472,  594, 476,  588, 1528,  588, 476,  592, 1524,  594, 470,  592, 476,  588, 476,  592, 1524,  594, 1524,  592, 1526,  592, 1524,  594, 472,  592, 1524,  592, 1524,  592, 1524,  592, 472,  592, 472,  592, 476,  592, 472,  592, 1524,  592, 472,  594};  // NEC 659A857A
uint32_t address = 0xA6;
uint32_t command = 0xA1;
uint64_t data = 0x659A857A;

SOURCE (long):
Encoding  : NEC
Code      : 659A05FA (32 bits)
Library   : v2.5.1

Raw Timing[135]:
   +  8552, -  4184,    +   596, -   472,    +   592, -  1524,    +   594, -  1524, 
   +   594, -   472,    +   592, -   472,    +   598, -  1520,    +   596, -   472, 
   +   594, -  1524,    +   592, -  1524,    +   592, -   472,    +   592, -   472, 
   +   596, -  1520,    +   598, -  1520,    +   596, -   472,    +   592, -  1524, 
   +   592, -   472,    +   592, -   476,    +   592, -   472,    +   592, -   472, 
   +   592, -   476,    +   592, -   472,    +   592, -  1524,    +   592, -   472, 
   +   598, -  1518,    +   598, -  1520,    +   596, -  1520,    +   596, -  1520, 
   +   596, -  1520,    +   596, -  1520,    +   596, -   472,    +   592, -  1524, 
   +   592, -   472,    +   598, - 25282,    +  8552, -  4182,    +   596, -  1520, 
   +   598, -  1518,    +   598, -  1520,    +   596, -  1520,    +   596, -   472, 
   +   592, -  1524,    +   592, -   472,    +   598, -  1520,    +   596, -   472, 
   +   594, -   472,    +   592, -   472,    +   596, -   472,    +   592, -  1524, 
   +   592, -   472,    +   592, -  1524,    +   596, -   472,    +   594, -  1520, 
   +   596, -  1520,    +   598, -   472,    +   592, -   472,    +   598, -   472, 
   +   594, -  1522,    +   594, -   472,    +   592, -  1524,    +   594, -   472, 
   +   596, -   472,    +   594, -  1524,    +   592, -  1524,    +   592, -  1524, 
   +   592, -   472,    +   594, -  1524,    +   598, -   472,    +   592

uint16_t rawData[135] = {8552, 4184,  596, 472,  592, 1524,  594, 1524,  594, 472,  592, 472,  598, 1520,  596, 472,  594, 1524,  592, 1524,  592, 472,  592, 472,  596, 1520,  598, 1520,  596, 472,  592, 1524,  592, 472,  592, 476,  592, 472,  592, 472,  592, 476,  592, 472,  592, 1524,  592, 472,  598, 1518,  598, 1520,  596, 1520,  596, 1520,  596, 1520,  596, 1520,  596, 472,  592, 1524,  592, 472,  598, 25282,  8552, 4182,  596, 1520,  598, 1518,  598, 1520,  596, 1520,  596, 472,  592, 1524,  592, 472,  598, 1520,  596, 472,  594, 472,  592, 472,  596, 472,  592, 1524,  592, 472,  592, 1524,  596, 472,  594, 1520,  596, 1520,  598, 472,  592, 472,  598, 472,  594, 1522,  594, 472,  592, 1524,  594, 472,  596, 472,  594, 1524,  592, 1524,  592, 1524,  592, 472,  594, 1524,  598, 472,  592};  // NEC 659A05FA
uint32_t address = 0xA6;
uint32_t command = 0xA0;
uint64_t data = 0x659A05FA;

crankyoldgit commented:
@kpalczewski Thanks for that. That's excellent data. Exactly what I needed/wanted.
I'll know in a few minutes when I build some test cases out of that data, but I'm fairly sure the data you've provided matches up with the data on the pioneer web page. i.e. The protocol is slightly/subtly different from what you had coded up.

crankyoldgit committed:
Correct Pioneer protocol based on real data.
* Remove inter-segment protocol message.
* Add better address/command decoding to encode/decode routines.
* Unit tests based on real example messages.

crankyoldgit commented:
@kpalczewski I was correct. We had the protocol slightly wrong. There is no "middle" bit. It's just two NEC messages back to back.

Anyway. I've pushed a commit that corrects that. If you test `IRrecvDumpV2` from your branch/fork after you `git pull` it should decode the long messages as Pioneer. The short ones will still decode as NEC, because that's what they are. The `encodePioneer()` function will also use the `address` and `command` values from that decode AND the short hex commands from the Pioneer's website and produce a 64bit Pioneer code for you.

crankyoldgit committed:
Merge branch 'master' into PioneerSendSupport

crankyoldgit commented:
As I've added significant code to this, I'll have someone else do a code review as well.

kpalczewski commented:
I tried to send Pioneer code:
  irsend.sendPioneer(0x659A857AF50A3DC2, 64, 0);

but the result wan't properly recognized.
The 2 NEC spacer is different:
+   592, - 40056, 
and it should be:
+ 598, - 25282, 

also, the second NEC codes are all 0s.


Here's the full output of the send code by esp8266:
Timestamp : 000015.029
Encoding  : NEC
Code      : 659A857A (32 bits)
Library   : v2.5.1

Raw Timing[135]:
   +  9018, -  4438,    +   622, -   510,    +   618, -  1628,    +   592, -  1654, 
   +   618, -   510,    +   618, -   506,    +   592, -  1652,    +   594, -   536, 
   +   594, -  1652,    +   592, -  1654,    +   594, -   532,    +   592, -   536, 
   +   618, -  1628,    +   592, -  1652,    +   594, -   536,    +   594, -  1654, 
   +   618, -   506,    +   596, -  1650,    +   598, -   532,    +   592, -   532, 
   +   594, -   536,    +   594, -   532,    +   592, -  1654,    +   592, -   538, 
   +   592, -  1654,    +   618, -   506,    +   592, -  1654,    +   592, -  1654, 
   +   596, -  1650,    +   596, -  1654,    +   594, -   532,    +   592, -  1654, 
   +   592, -   540,    +   592, - 40056,    +  9014, -  4434,    +   596, -   532, 
   +   618, -   506,    +   596, -   532,    +   596, -   528,    +   592, -   532, 
   +   618, -   506,    +   596, -   532,    +   618, -   506,    +   592, -   538, 
   +   592, -   532,    +   594, -   536,    +   594, -   532,    +   592, -   532, 
   +   592, -   536,    +   620, -   506,    +   594, -   536,    +   620, -   506, 
   +   598, -   532,    +   592, -   532,    +   594, -   536,    +   594, -   532, 
   +   618, -   510,    +   594, -   530,    +   592, -   532,    +   592, -   536, 
   +   592, -   532,    +   618, -   512,    +   592, -   532,    +   618, -   510, 
   +   592, -   532,    +   592, -   532,    +   596, -   532,    +   592

uint16_t rawData[135] = {9018, 4438,  622, 510,  618, 1628,  592, 1654,  618, 510,  618, 506,  592, 1652,  594, 536,  594, 1652,  592, 1654,  594, 532,  592, 536,  618, 1628,  592, 1652,  594, 536,  594, 1654,  618, 506,  596, 1650,  598, 532,  592, 532,  594, 536,  594, 532,  592, 1654,  592, 538,  592, 1654,  618, 506,  592, 1654,  592, 1654,  596, 1650,  596, 1654,  594, 532,  592, 1654,  592, 540,  592, 40056,  9014, 4434,  596, 532,  618, 506,  596, 532,  596, 528,  592, 532,  618, 506,  596, 532,  618, 506,  592, 538,  592, 532,  594, 536,  594, 532,  592, 532,  592, 536,  620, 506,  594, 536,  620, 506,  598, 532,  592, 532,  594, 536,  594, 532,  618, 510,  594, 530,  592, 532,  592, 536,  592, 532,  618, 512,  592, 532,  618, 510,  592, 532,  592, 532,  596, 532,  592};  // NEC 659A857A
uint32_t address = 0xA6;
uint32_t command = 0xA1;
uint64_t data = 0x659A857A;



when I use my remote, it's properly recognized:
Timestamp : 000203.041
Encoding  : PIONEER
Code      : 659A857AF50A3DC2 (64 bits)
Library   : v2.5.1

Raw Timing[135]:
   +  8556, -  4180,    +   598, -   472,    +   594, -  1524,    +   592, -  1520, 
   +   596, -   472,    +   592, -   472,    +   594, -  1524,    +   592, -   472, 
   +   592, -  1524,    +   592, -  1524,    +   594, -   472,    +   594, -   476, 
   +   592, -  1522,    +   596, -  1520,    +   592, -   476,    +   594, -  1520, 
   +   596, -   472,    +   592, -  1524,    +   592, -   472,    +   592, -   472, 
   +   594, -   476,    +   588, -   476,    +   592, -  1524,    +   594, -   472, 
   +   592, -  1524,    +   594, -   472,    +   594, -  1524,    +   594, -  1524, 
   +   592, -  1524,    +   594, -  1524,    +   594, -   472,    +   592, -  1524, 
   +   592, -   472,    +   594, - 25286,    +  8556, -  4186,    +   588, -  1524, 
   +   594, -  1524,    +   594, -  1524,    +   592, -  1524,    +   594, -   476, 
   +   588, -  1524,    +   594, -   476,    +   592, -  1524,    +   594, -   472, 
   +   592, -   472,    +   594, -   474,    +   588, -   476,    +   592, -  1524, 
   +   594, -   472,    +   592, -  1524,    +   594, -   470,    +   594, -   472, 
   +   592, -   476,    +   588, -  1524,    +   592, -  1524,    +   594, -  1524, 
   +   592, -  1524,    +   592, -   476,    +   594, -  1520,    +   594, -  1524, 
   +   592, -  1524,    +   594, -   476,    +   594, -   472,    +   592, -   472, 
   +   592, -   476,    +   588, -  1524,    +   592, -   476,    +   592

uint16_t rawData[135] = {8556, 4180,  598, 472,  594, 1524,  592, 1520,  596, 472,  592, 472,  594, 1524,  592, 472,  592, 1524,  592, 1524,  594, 472,  594, 476,  592, 1522,  596, 1520,  592, 476,  594, 1520,  596, 472,  592, 1524,  592, 472,  592, 472,  594, 476,  588, 476,  592, 1524,  594, 472,  592, 1524,  594, 472,  594, 1524,  594, 1524,  592, 1524,  594, 1524,  594, 472,  592, 1524,  592, 472,  594, 25286,  8556, 4186,  588, 1524,  594, 1524,  594, 1524,  592, 1524,  594, 476,  588, 1524,  594, 476,  592, 1524,  594, 472,  592, 472,  594, 474,  588, 476,  592, 1524,  594, 472,  592, 1524,  594, 470,  594, 472,  592, 476,  588, 1524,  592, 1524,  594, 1524,  592, 1524,  592, 476,  594, 1520,  594, 1524,  592, 1524,  594, 476,  594, 472,  592, 472,  592, 476,  588, 1524,  592, 476,  592};  // PIONEER 659A857AF50A3DC2
uint32_t address = 0xA6A1;
uint32_t command = 0xAFBC;
uint64_t data = 0x659A857AF50A3DC2;

crankyoldgit commented:
@kpalczewski Thanks for that feedback. I'll take a look and try to see what might be causing it to misbehave.

crankyoldgit committed:
Additional unit test for Pioneer protocol.
Trying to debug:
https://github.com/markszabo/IRremoteESP8266/pull/547#issuecomment-430800734

crankyoldgit commented:
@kpalczewski I tried the `irsend.sendPioneer(0x659A857AF50A3DC2, 64, 0);` you suggested, and my testing (via unit tests only) seems to indicate it is working as expected. i.e. The second half isn't all zeros.
You can see that clearly here:
https://github.com/markszabo/IRremoteESP8266/pull/547/commits/545089a0a7276675eb72dfc10d9a1273a4ab5d49#diff-e04812f13f3ef2304b3e794d63eec982R131

Perhaps try _irsend.sendPioneer(0x659A857AF50A3DC2**ULL**, 64, 0);_ instead? That should ensure the compiler sees the hex code as a 64 bit value. `ULL` means _unsigned long long_. Note: You shouldn't need to specify the `,64, 0` bit either. I would have expected the first half of the message to be zeros, not the second if it was using a 32bit value instead of 64bit. So I don't expect it to work. But it's worth a shot.

Maybe wipe all the `*.o` files and compile/upload again? Or can include the entire source code for the simplest send example that causes it?

kpalczewski commented:
@crankyoldgit  Apparently Arduino IDE 1.8.6 on Windows 7 compiles this line differently:
`sendNEC(data & ((1UL << (nbits / 2)) - 1), nbits / 2, repeat);`
this sends 0, because the value calculated is 0.

`sendNEC(data & ((0xFFFFFFFF, nbits / 2, repeat);`
this sends proper code. I guess 0xFF isn't very elegant.

Even though the space between the codes is 40056 rather than 25286 (value from the remote), the Pioneer receiver properly recognizes the codes.

crankyoldgit commented:
Yep. I see my mistake now. And why the unit tests pass and why it fails on the arduino.
Fix coming soon.

crankyoldgit committed:
Fix an arduino integer size issue.
i.e. UL on x86/linux is 64 bit. On Arduino, it's 32bit. Force 64bit.

crankyoldgit commented:
@kpalczewski Can you try it after that latest patch? That should fix it. Thanks again for finding the root cause of the problem.

crankyoldgit committed:
Merge branch 'master' into PioneerSendSupport

kpalczewski commented:
the pleasure is all mine.
The code works, all NEC codes are send properly, the receiver recognizes them.

crankyoldgit commented:
Excellent. Assuming no further problems this will be merged in to the master branch within 2 days.
I'll probably make a new official release shortly after that barring any emergencies.

crankyoldgit commented:
It's 7 days since I requested a review by someone else. No negative feedback. So merging.

crankyoldgit commented:
This is now live in the new _v2.5.2_ release of the library.
